File: Coder Army/SOLID/DIP(dependencyInversionPrinciple)/DIP_followed.cpp
#include <bits/stdc++.h>
using namespace std;

class Database
{
public:
    virtual void save(string data) = 0;
};

class MySQLDatabase : public Database
{
public:
    void save(string data) override
    {
        cout << "Insert into db.company.users value ('" << data << "')" << endl;
    }
};
class MongodbDatabase : public Database
{
public:
    void save(string data) override
    {
        cout << "Insert into MongodbDatabase value ('" << data << "')" << endl;
    }
};

class CassandraDatabase : public Database
{
public:
    void save(string data) override
    {
        cout << "Insert into CassandraDatabase value ('" << data << "')" << endl;
    }
};

class UserServices
{
private:
    Database *db; // dependency injection

public:
    UserServices(Database *db) : db(db) {}

    void storeToDB(string user)
    {
        db->save(user);
    }
};
int main()
{

    MySQLDatabase mySQL;
    MongodbDatabase mongodb;
    CassandraDatabase cdb;

    UserServices service(&mySQL);
    service.storeToDB("Kuldeep");

    return 0;
}

File: Coder Army/SOLID/DIP(dependencyInversionPrinciple)/Javascript code/DIP_followed.js
class Database {
  save(data) {
    throw new Error("Method 'save()' must be implemented.");
  }
}

class MySQLDatabase extends Database {
  save(data) {
    console.log(`Insert into db.company.users value ('${data}')`);
  }
}

class MongodbDatabase extends Database {
  save(data) {
    console.log(`Insert into MongodbDatabase value ('${data}')`);
  }
}

class CassandraDatabase extends Database {
  save(data) {
    console.log(`Insert into CassandraDatabase value ('${data}')`);
  }
}

class UserServices {
  constructor(db) {
    this.db = db; // Dependency injection
  }

  storeToDB(user) {
    this.db.save(user);
  }
}

// Usage
const mySQL = new MySQLDatabase();
const mongoDB = new MongodbDatabase();
const cassandraDB = new CassandraDatabase();

const service = new UserServices(mySQL);
service.storeToDB("Kuldeep");


const mongoService = new UserServices(mongoDB);
mongoService.storeToDB("Amit");

const cassandraService = new UserServices(cassandraDB);
cassandraService.storeToDB("Ravi");
